### PR TITLE
Miscellaneous fixes/improvements

### PR DESCRIPTION
**Pull request by colinxs**
As promised @oschulz, here's the PR! Everything should be non-breaking (I didn't break any tests at least). The only "questionable" change is 2a562cf which adds `sizehint!`/`resize!` where only the size in the last dimension is specified. Given that this is the only resizeable dimension it seems useful, but perhaps it would make more sense call them something different e.g. `sizehintlastdim!`/`resizelastdim`?

### CODE REVIEW

**Comment by colinxs** [comment]
Once this is reviewed/modified/approved I'll make the PR with broadcasting support :)

**Comment by colinxs** [comment]
If you're wondering why I ditched `_copyto_impl`: the explicit version appeared to generate simpler LLVM code and it matched the style in `Base`.

File: src/elasticarray.jl
Comment (oschulz) [inline src/elasticarray.jl]: That seems to break the semantics of `append!`, for example

```julia
append!(rand(2), (rand(1, 2) for i in 1:4))
```

is not allowed. `append!` should append the contents of `src`, not of each element of `src` iteratively. The `append!` implemented here would be suitable for an `ArrayOfSimilarArrays` wrapped around an `ElasticArray`.

File: src/elasticarray.jl
Comment (oschulz) [inline src/elasticarray.jl:124]: The coding style of `ElasticArrays` actually doesn't use `return`. It's a matter of taste though, I'll admit.

File: src/elasticarray.jl
Comment (oschulz) [inline src/elasticarray.jl]: See comment on `append!` above.

File: src/elasticarray.jl
Comment (oschulz) [inline src/elasticarray.jl:155]: Please use `function ... end` instead of `(... ; ...)`.

File: src/elasticarray.jl
Comment (oschulz) [inline src/elasticarray.jl:178]: Just curious, did you add `@inline` just in case (I'm fine with that!), or did you already observe problems with inlining here?

File: src/elasticarray.jl
Comment (oschulz) [inline src/elasticarray.jl]: I thought about allowing this a while back, but decided against it - in principle, one could also add an array type that is elastic in other dimensions (see #1), and then having a `resize` that takes only one number for a multi-dim array would be a bit unclean.

However, we could export a `resize_lastdim!` function - doesn't seem to be claimed by any package yet.

File: src/elasticarray.jl
Comment (oschulz) [inline src/elasticarray.jl]: See comment on `resize!` with single argument. We could export a `sizehint_lastdim!` function, though.

File: src/elasticarray.jl
Comment (oschulz) [inline src/elasticarray.jl:96]: I'd like to stay with `dims::Vararg{Integer,N}` instead of `dims::NTuple{N,Integer}`, it's more in line with Julia array ctor syntax.

File: src/elasticarray.jl
Comment (oschulz) [inline src/elasticarray.jl:82]: I'd like to stay with `dims::Vararg{Integer,N}` instead of `dims::NTuple{N,Integer}`, it's more in line with Julia array ctor syntax.

**Review by oschulz — CHANGES_REQUESTED** [review]

**Comment by oschulz** [comment]
> If you're wondering why I ditched _copyto_impl

Thanks for the explanation, I actually had wondered about that. :-)

File: src/elasticarray.jl
Comment (colinxs) [reply src/elasticarray.jl:82]: Both `dims::Vararg{Integer,N}` and `dims::NTuple{N,Integer}` are supported. Given that `resize!`/`sizehint!` are only defined for 1-D array types (Vector, BitVector, etc), it doesn't seem like there's precedence here? 

I added both since most other `Base` functions that have `dims` in the function signature allow `dims` to be specified as either `Varrg` or `NTuple`:
```
Array{Int}(undef, 1, 2)
Array{Int, undef, (1, 2))
reshape(A, 1, 2)
reshape(A, (1, 2))
```
Are all valid.

File: src/elasticarray.jl
Comment (colinxs) [reply src/elasticarray.jl:96]: see below

File: src/elasticarray.jl
Comment (colinxs) [reply src/elasticarray.jl]: sounds good!

File: src/elasticarray.jl
Comment (colinxs) [reply src/elasticarray.jl]: sounds good!

File: src/elasticarray.jl
Comment (colinxs) [reply src/elasticarray.jl:178]: As a rule-of-thumb I add `@inline` if the function in `Base` is also defined with `@inline`/`@_inline_meta` (e.g. `Base.pointer`), but in this case it's not! I can remove it.

File: src/elasticarray.jl
Comment (colinxs) [reply src/elasticarray.jl:155]: sounds good!
In regards to this and your above comment on style, what style do you use? I've been following BlueStyle.

File: src/elasticarray.jl
Comment (colinxs) [reply src/elasticarray.jl:124]: sounds good (see below comment on style)

File: src/elasticarray.jl
Comment (oschulz) [reply src/elasticarray.jl:155]: Uhm, my own, kinda. :-) I just don't use `return` much (except where really necessary) to keep the "functional programming" feel of Julia.

File: src/elasticarray.jl
Comment (oschulz) [reply src/elasticarray.jl:178]: No, no, I'm all for using explicit `@inline`s, I seen so many cases where you'd think "the compiler will surely inline this", and then it doesn't. I was just curious if you'd seen something like that here.

File: src/elasticarray.jl
Comment (oschulz) [reply src/elasticarray.jl:82]: Good point, yes!

File: src/elasticarray.jl
Comment (colinxs) [reply src/elasticarray.jl]: I see your point. It would still be nice to support appending/prepending from any iterable to be consistent with `Base` (e.g. `append!(ElasticArray{Float64}(undef, 2, 3), (rand() for _=1:2))`). If we changed this to: 
```julia
function Base.append!(dest::ElasticArray, iter)
    append!(dest.data, iter)
    # check rem(length(eachindex(A.data)), A.kernel_length) == 0
    return dest
end
 ```
then it would be consistent with the semantics of `append!`, correct? Since we don't have access to `length` for an arbitrary iterator we'd have to check if `rem(length(eachindex(A.data)), A.kernel_length) == 0` after appending/prepending.

File: src/elasticarray.jl
Comment (colinxs) [reply src/elasticarray.jl:178]: For this particular case I haven't :), but in general I error on the side of inlining haha

File: src/elasticarray.jl
Comment (oschulz) [reply src/elasticarray.jl]: Sounds good to  me!

File: src/elasticarray.jl
Comment (oschulz) [reply src/elasticarray.jl:178]: Me too, go right ahead. :-)

File: src/elasticarray.jl
Comment (colinxs) [reply src/elasticarray.jl]: See https://github.com/JuliaArrays/ElasticArrays.jl/pull/20/commits/fc1f4e7d8fa9b33c6df664596e65086c8df1b2fe. Turns out we don't need a separate `prepend!`/`append!` when `src` is an `AbstractArray` :).

If that looks good to you feel free to resolve this and your comment on `prepend!`

File: src/elasticarray.jl
Comment (colinxs) [reply src/elasticarray.jl:155]: Yeah the most important thing about style is just that its consistent haha. Would you be open to adopting one of the more popular styles like BlueStyle? There's good tooling for auto-formatting and it's always nice to see similar styles across packages. (although you won't like their suggestion for `return` which is to use it for multi-line functions specified as `function ... end` :) )

Comment (oschulz) [reply src/elasticarray.jl:155]: Yes, I'm open to BlueStyle - have to admit, I never looked at it before, but Lyndon & friends at Invenia will certainly know what they are doing, and they have a large code base to maintain.

File: src/elasticarray.jl
Comment (oschulz) [reply src/elasticarray.jl:124]: Sorry, did accedentially mark this as resolved.

Comment (oschulz) [reply src/elasticarray.jl:124]: So we keep the `return`s. :-)

File: src/elasticarray.jl
Comment (oschulz) [reply src/elasticarray.jl]: Sorry, did accidentally mark this as resolved.

File: src/elasticarray.jl
Comment (oschulz) [reply src/elasticarray.jl]: Sorry, did accidentally mark this as resolved.

File: src/elasticarray.jl
Comment (colinxs) [reply src/elasticarray.jl]: Perhaps we can document it and leave it unexported? That gives us some flexibility in the future.

Comment (colinxs) [reply src/elasticarray.jl]: @oschulz should be fixed now

File: src/elasticarray.jl
Comment (colinxs) [reply src/elasticarray.jl]: @oschulz should be fixed now

File: src/elasticarray.jl
Comment (oschulz) [reply src/elasticarray.jl]: Yes, maybe better not not export it for now. If we find out it's frequently used, we can still export it later on (the function name seems to be available).

File: src/elasticarray.jl
Comment (oschulz) [reply src/elasticarray.jl]: Thanks!

**Comment by oschulz** [comment]
@colinxs , I'll wait for your PR before tagging a new release, Ok? Then we can decide whether it'll be v1.1 or v2.0.

**Comment by oschulz** [comment]
@colinxs , I'll wait for your PR with the broadcasting support before tagging a new release, Ok? Then we can decide whether it'll be v1.1 or v2.0.

**Comment by colinxs** [comment]
@oschulz sounds good!

**Pull request merged**
